docs/libcudacxx/extended_api/execution_model.rst

```diff
@@ -0,0 +1,261 @@
+    using "host.threads.<id>" and "device.threads.<id>", respectively.
+
+    .. code:: cuda
+        // Example: Execution.Model.Device.0
+        // Outcome: grid eventually terminates per device.threads.4 because the atomic object does not have automatic storage duration.
+        __global__ void ex0(cuda::atomic_ref<int, cuda::thread_scope_device> atom) {
+            if (threadIdx.x == 0) {
+                while(atom.load(cuda::memory_order_relaxed) == 0);
+            } else if (threadIdx.x == 1) {
+                atom.store(1, cuda::memory_order_relaxed);
+            }
+        }
+
+    .. code:: cuda
+        // Example: Execution.Model.Device.1
+        // Allowed outcome: No thread makes progress because device threads don't support host.threads.2.
+        __global__ void ex1() {
+            while(true) std::this_thread::yield();
+        }
+
+    .. code:: cuda
+        // Example: Execution.Model.Device.2
+        // Allowed outcome: No thread makes progress because device threads don't support host.threads.4
+        // for objects with automatic storage duration (see exception in device.threads.3).
+        __global__ void ex2() {
+            volatile bool True = true;
+            while(True);
+        }
+
+    .. code:: cuda
+        // Example: Execution.Model.Device.3
+        // Allowed outcome: No thread makes progress because device threads don't support host.threads.5
+        // for objects with automatic storage duration (see exception in device.threads.4).
+        __global__ void ex3() {
+            cuda::atomic<bool, cuda::thread_scope_thread> True = true;
+            while(True.load());
+        }
+
+    .. code:: cuda
+        // Example: Execution.Model.Device.4
+        // Allowed outcome: No thread makes progress because device threads don't support host.thread.6.
+        __global void ex4() {
+            while(true) { /* empty */ }
+        }
+
+.. _libcudacxx-extended-api-execution-model-cuda-apis:
+
+CUDA APIs
+---------
+
+Any CUDA API shall eventually either return or ensure at least one device thread makes progress.
+
+CUDA query functions (e.g. `cudaStreamQuery <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g2021adeb17905c7ec2a3c1bf125c5435>`__,
+`cudaEventQuery <https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g2bf738909b4a059023537eaa29d8a5b7>`__, etc.) shall not consistently
+return ``cudaErrorNotReady`` without a device thread making progress.
+
+[Note: The device thread need not be "related" to the API call, e.g., an API operating on one stream or process may ensure progress of a device thread on another stream or process. - end note.]
+
+[Note: A simple but not sufficient method to test workloads for CUDA API Forward Progress conformance is to run them with following environment variables set: ``CUDA_DEVICE_MAX_CONNECTIONS=1 CUDA_LAUNCH_BLOCKING=1`` - end note.]
```

Comment:
if it is insufficient, why mention it here?

Reply:
How do I test my program conforms to forward progress is a frequently-asked question by users.

Testing as suggested is our recommended way to do that. While insufficient, it will catch many/most issues, and is the only "tool" we provide for this. It is therefore worth documenting somewhere, and for now, this document is the only place in our entire documentation in which we talk about this topic, so "here" seemed better than "nowhere".

If we eventually develop high-level user-documentation for any of this, we should probably expand on this there.

Reply:
Does the following make it clearer?

```suggestion
[Note: A simple but not sufficient method to test a program for CUDA API Forward Progress conformance is to run it with following environment variables set: ``CUDA_DEVICE_MAX_CONNECTIONS=1 CUDA_LAUNCH_BLOCKING=1`` and check that the program still terminates. 
If it does not, the program may have a Forward Progress bug.
This method, while insufficient, does catch many such Forward Progress bugs. - end note.]
```

Reply:
@ericniebler resolved, please re-open if this does not address the issue raised.